Add structured event for open redirects

adrianna-chang-shopify opened the pull request:
### Motivation / Background

Follow up to: https://github.com/rails/rails/pull/55496

This Pull Request has been created so that applications with `config.action_controller.action_on_open_redirect = :notify` automatically get a structured log event, rather than having to configure their own subscriber to the Active Support Notification.

### Detail

Adds a method for `open_redirect` to `ActionController::StructuredEventSubscriber`. Uses the last line of the stacktrace for `stack_trace` to match `#rescue_from_callback`, but we could offer a bit more here if desired (return the last X lines, filter the trace, etc.)

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.

skipkayhil commented on actionpack/lib/action_controller/metal/redirecting.rb:
```diff
@@ -292,7 +292,7 @@ def _enforce_open_redirect_protection(location, allow_other_host:)
             location: location,
             request: request,
-            stack_trace: caller,
+            stacktrace: caller,
```
This was released in 8.1, so I'm not sure its worth breaking backwards compatibility to remove the underscore

gmcgibbon replied:
We can remove the underscore in the new structured event subscriber method, but not here. The notification must remain the same.

skipkayhil committed:
Add structured event for open redirects

Adds a call to the structured event reporter for "action_controller.open_redirect_detected".
Applications that have `config.action_controller.action_on_open_redirect = :notify`
will automatically get a structured log.

Co-authored-by: Adrianna Chang <adrianna.chang@shopify.com>

hmcguire-shopify commented:
I pushed up new commits to split out the backportable changes from the new feature. (I also dropped the change in log_subscriber because https://github.com/rails/rails/pull/56059 makes it unnecessary)

However, now that I've spent more time with the code I don't think we should be filtering the `stacktrace` in the event payload (related: https://github.com/rails/rails/issues/56060). We are more frequently interested in frames further up the stack than just the first one, and we lose that information by filtering here.

gmcgibbon reviewed (approved):
LGTM, but I think we should still add a changelog for this. The structured event subscriber is released in 8.1.
